File: adv15.py
#!/usr/bin/env python3

factorA = 16807
factorB = 48271
base = 2147483647
A = 699
B = 124

if False:
    A = 65
    B = 8921

def step(start, factor, mul):
    result = start * factor % base
    while result % mul:
        result = result * factor % base
    return result

result = 0
for _ in range(5000000):
    A, B = step(A, factorA, 4), step(B, factorB, 8)
    if A % 65536 == B % 65536:
        result += 1

print(result)

File: 2017/adv15.c
long base = 2147483647;
unsigned long step(unsigned long start, unsigned int fac, unsigned int mul) {
  do {
    start = start * fac % base;
  } while (start & mul);
  return start;
}

int main() {
  long a = 699;
  long b = 124;
  unsigned int factorA = 16807;
  unsigned int factorB = 48271;
  long i;
  int result = 0;
  for (i = 0; i < 5000000; i++) {
    a = step(a, factorA, 3U);
    b = step(b, factorB, 7U);
    result += !((a ^ b) & 65535);
  }
  printf("%d\n", result);
}
